### PR TITLE
NXP backend: Add support for optimizing Conv+BN during QAT

### PR DESCRIPTION
**Pull request by StrycekSimon**
### Summary
Enables optimization of Conv+BatchNorm during QAT. This involves:
- Enabling TorchAO native batch norm and conv fusing (by disabling our `FuseBatchNormWithConvPass` when in QAT mode)
- Removing output quantization of convolution (for native conv+bn fusing to properly match the pattern to be replaced)
- Adding BatchNorm quantization output pattern implementation

### Test plan
Test cases for Conv+BatchNorm fusion and quantization were added.

cc @robert-kalmar @JakeStevens @digantdesai

### CODE REVIEW

File: backends/nxp/tests/test_quantizer.py [hunk 1]
```diff
@@ -636,3 +636,50 @@ def test_qat_produces_same_graph_as_ptq():
             qat_quantized_model.graph.nodes, ptq_quantized_model.graph.nodes
         )
     )
+
+
+@pytest.mark.parametrize("conv_module", ["conv1d", "conv2d", "conv2d_t"])
```
[the remark this answers is not in the document]
Comment (StrycekSimon) [reply backends/nxp/tests/test_quantizer.py:650]: The `conv1d_t` is currently unsupported. I added a comment stating this fact.

File: backends/nxp/quantizer/neutron_quantizer.py [hunk 2]
```diff
@@ -293,7 +298,16 @@ def transform_for_annotation(
...
 
-        model = NeutronAtenPassManager(self.neutron_target_spec)(model).graph_module
+        pass_manager = NeutronAtenPassManager(self.neutron_target_spec)
+
+        if self.is_qat:
+            pass_manager.passes = [
+                p
+                for p in pass_manager.passes
+                if not isinstance(p, FuseBatchNormWithConvPass)
+            ]
```
[the remark this answers is not in the document]
Comment (StrycekSimon) [reply backends/nxp/quantizer/neutron_quantizer.py]: I refactored it and moved the comment to the `neutron_aten_pass_manager.py` file.

File: backends/nxp/tests/test_quantizer.py [hunk 3]
```diff
@@ -636,3 +636,51 @@ def test_qat_produces_same_graph_as_ptq():
...
+
+
+# TODO: conv1d_t is currently unsupported, add when resolved
+@pytest.mark.parametrize("conv_module", ["conv1d", "conv2d", "conv2d_t"])
+@pytest.mark.parametrize("conv_bias", [True, False])
+@pytest.mark.parametrize("bn_affine", [True, False])
+def test_torchao_native_conv_bn_qat_fusing(conv_module, conv_bias, bn_affine):
+    if not conv_bias:
+        pytest.skip("Conv without bias is not supported.")
+
```
[the remark this answers is not in the document]
Comment (StrycekSimon) [reply backends/nxp/tests/test_quantizer.py:656]: I am keeping it as it is a subject of future enablement.

File: examples/nxp/aot_neutron_compile.py [hunk 4]
```diff
@@ -149,6 +149,13 @@ def get_model_and_inputs_from_name(model_name: str):
         default=False,
         help="Produce a quantized model",
     )
+    parser.add_argument(
+        "--use_qat",
+        action="store_true",
+        required=False,
+        default=False,
+        help="Use QAT mode for quantization (does not include QAT training)",
```
Comment (MartinPavella) [inline examples/nxp/aot_neutron_compile.py]: If the quantization aware **training** is not possible using this module, why include it? Just to show how it can be triggered? If so, perhaps a separate example module, or even just a README might be better in my opinion.

**Review by MartinPavella — APPROVED** [review]

**Comment by StrycekSimon** [comment]
The failing unittest tests are related to XNNPack backend. Although I have not seen such error in other PRs, implementation in this PR should not interfere with it.

File: examples/nxp/aot_neutron_compile.py (hunk 4 above)
Comment (StrycekSimon) [reply examples/nxp/aot_neutron_compile.py]: Good point, I will add QAT training at least for the CifarNet example.

File: examples/nxp/experimental/cifar_net/cifar_net.py [hunk 5]
```diff
@@ -145,35 +145,12 @@ def get_model(
...
-                running_loss += loss.item()
-                if i % 2000 == 1999:  # print every 2000 mini-batches
-                    print(f"[{epoch + 1}, {i + 1:5d}] loss: {running_loss / 2000:.3f}")
-                    running_loss = 0.0
-
-        logger.info("Finished training.")
-        if state_dict_file is not None and train:
-            logger.info(f"Saving the trained weights in `{state_dict_file}`.")
+        cifar_net = train_cifarnet_model(
+            model=cifar_net, batch_size=batch_size, num_epochs=num_epochs
```
[the remark this answers is not in the document]
Comment (StrycekSimon) [reply examples/nxp/experimental/cifar_net/cifar_net.py]: You are right, I will update it...

File: examples/nxp/experimental/cifar_net/cifar_net.py [hunk 6]
```diff
@@ -189,6 +166,40 @@ def get_cifarnet_calibration_data(num_images: int = 100) -> tuple[torch.Tensor]:
     return (tensor,)
 
 
+def train_cifarnet_model(
+    cifar_net: nn.Module | torch.fx.GraphModule,
+    batch_size: int = 1,
+    num_epochs: int = 1,
+) -> nn.Module:
```
[the remark this answers is not in the document]
Comment (StrycekSimon) [reply examples/nxp/experimental/cifar_net/cifar_net.py:173]: I will keep it named `cifar_net` for consistency with `test_cifarnet_model` function.

File: backends/nxp/quantizer/patterns.py [hunk 7]
```diff
@@ -524,11 +570,20 @@ def get_anchors(
...
+            if possibly_bn and _is_batch_norm(possibly_bn):
+                output_specs = []
+
         return PartitionAnchors(
             inputs=[(conv_node, NodeArgsIdx(0))],
             weights=[(conv_node, NodeArgsIdx(1), weight_quantization_spec)],
             biases=bias,
-            output=[(conv_node,)],
+            output=output_specs,
         )
```
[the remark this answers is not in the document]
Comment (StrycekSimon) [reply backends/nxp/quantizer/patterns.py:587]: Good one, I missed it in my previous QAT-related PR. There is a deeper problem that needs to be resolved before adding it. I will log it and ship this change in a separate PR.

File: examples/nxp/aot_neutron_compile.py (hunk 4 above)
Comment (StrycekSimon) [reply examples/nxp/aot_neutron_compile.py]: I added QAT training support for CifarNet model.

File: backends/nxp/tests/test_quantizer.py [hunk 8]
```diff
@@ -636,3 +636,51 @@ def test_qat_produces_same_graph_as_ptq():
...
+    if len(conv_node_args) > 3:
+        conv_node_args = conv_node_args[:3]
+
+    assert len([n for n in graph_nodes if "batch_norm" in n.name]) == 0
+    assert (
+        len(conv_node.users) == 1
+        and list(conv_node.users.keys())[0].target
+        == torch.ops.quantized_decomposed.quantize_per_tensor.default
+    )
+    assert all(arg.name.startswith("dequantize") for arg in conv_node_args)
```
Comment (roman-janik-nxp) [inline backends/nxp/tests/test_quantizer.py]: Please use target instead of names, we did changes from name to target before.

File: backends/nxp/tests/test_quantizer.py [hunk 9]
```diff
@@ -636,3 +636,51 @@ def test_qat_produces_same_graph_as_ptq():
...
+        ]
+
+    graph_nodes = list(quantized_model.graph.nodes)
+    conv_node = next(n for n in graph_nodes if is_conv(n))
+    conv_node_args = conv_node.args
+
+    if len(conv_node_args) > 3:
+        conv_node_args = conv_node_args[:3]
+
+    assert len([n for n in graph_nodes if "batch_norm" in n.name]) == 0
```
Comment (roman-janik-nxp) [inline backends/nxp/tests/test_quantizer.py]: Please use target instead of names, we did changes from name to target before.

File: backends/nxp/quantizer/neutron_quantizer.py
Comment (roman-janik-nxp) [inline backends/nxp/quantizer/neutron_quantizer.py]: 2026

File: backends/nxp/quantizer/patterns.py
Comment (roman-janik-nxp) [inline backends/nxp/quantizer/patterns.py]: 2025-2026

File: backends/nxp/aten_passes/neutron_aten_pass_manager.py
Comment (roman-janik-nxp) [inline backends/nxp/aten_passes/neutron_aten_pass_manager.py]: 2025-2026

File: examples/nxp/aot_neutron_compile.py
Comment (roman-janik-nxp) [inline examples/nxp/aot_neutron_compile.py]: 2024-2026

**Review by roman-janik-nxp — APPROVED** [review]

**Review by jirioc — APPROVED** [review]

**Comment by roman-janik-nxp** [comment]
The failing tests are not caused by this PR.
